File: expo-lan-sockets/android/src/main/java/com/opengamesonline/expolansockets/ExpoLanSocketsModule.kt
package com.opengamesonline.expolansockets

import android.content.Context
import android.net.nsd.NsdManager
import android.net.nsd.NsdServiceInfo
import android.net.wifi.WifiManager
import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record
import java.net.InetSocketAddress
import java.net.InetAddress
import java.net.ServerSocket
import java.net.Socket
import java.net.SocketException
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.Executors
import java.util.concurrent.atomic.AtomicBoolean

private const val MAX_CONNECTIONS = 8
private const val CONNECT_TIMEOUT_MS = 10_000

internal data class ServerOptions(
  @Field val serviceName: String = "Expo LAN Game",
  @Field val serviceType: String = "_expo-lan-game._tcp."
) : Record

private class ManagedConnection(val id: String, val socket: Socket, val incoming: Boolean) {
  val writeLock = Any()
  val closed = AtomicBoolean(false)
}

class ExpoLanSocketsModule : Module() {
  private val stateLock = Any()
  private val executor = Executors.newCachedThreadPool()
  private val connections = ConcurrentHashMap<String, ManagedConnection>()
  private val discoveredServices = ConcurrentHashMap<String, NsdServiceInfo>()
  private val discoveredServiceIds = ConcurrentHashMap<String, String>()

  private var serverSocket: ServerSocket? = null
  private var registrationListener: NsdManager.RegistrationListener? = null
  private var serverStopPromise: Promise? = null
  private var discoveryListener: NsdManager.DiscoveryListener? = null
  private var multicastLock: WifiManager.MulticastLock? = null
  private var multicastUsers = 0
  private var destroyed = false

  private val nsdManager: NsdManager
    get() = requireNotNull(appContext.reactContext)
      .getSystemService(Context.NSD_SERVICE) as NsdManager

  override fun definition() = ModuleDefinition {
    Name("ExpoLanSockets")

    Events(
      "onServiceFound",
      "onServiceLost",
      "onConnectionOpened",
      "onMessage",
      "onConnectionClosed",
      "onError"
    )

    Constant("capabilities") {
      mapOf(
        "tcpServer" to true,
        "tcpClient" to true,
        "serviceDiscovery" to true
      )
    }

    AsyncFunction("startServerAsync") { options: ServerOptions, promise: Promise ->
      startServer(options, promise)
    }

    AsyncFunction("stopServerAsync") { promise: Promise ->
      stopServer(promise)
    }

    AsyncFunction("startDiscoveryAsync") { serviceType: String, promise: Promise ->
      startDiscovery(serviceType, promise)
    }

    AsyncFunction("stopDiscoveryAsync") { promise: Promise ->
      stopDiscovery(promise)
    }

    AsyncFunction("connectAsync") { host: String, port: Int, promise: Promise ->
      executor.execute { connect(host, port, promise) }
    }

    AsyncFunction("connectToServiceAsync") { serviceId: String, promise: Promise ->
      connectToService(serviceId, promise)
    }

    AsyncFunction("sendAsync") { connectionId: String, data: ByteArray ->
      val connection = connections[connectionId]
        ?: throw LanSocketsException("ERR_CONNECTION_NOT_FOUND", "Connection '$connectionId' was not found")
      write(connection, data)
    }

    AsyncFunction("broadcastAsync") { data: ByteArray ->
      connections.values.forEach { connection -> write(connection, data) }
    }

    AsyncFunction("disconnectAsync") { connectionId: String ->
      connections[connectionId]?.let { closeConnection(it, "local_close") }
    }

    OnDestroy {
      destroyed = true
      stopDiscovery(null)
      stopServer(null)
      executor.shutdownNow()
    }
  }

  private fun startServer(options: ServerOptions, promise: Promise) {
    synchronized(stateLock) {
      if (serverSocket != null) {
        promise.reject("ERR_SERVER_ALREADY_RUNNING", "A server is already running", null)
        return
      }
    }

    executor.execute {
      val server = try {
        ServerSocket(0)
      } catch (error: Exception) {
        promise.reject("ERR_SERVER_START", error.message ?: "Could not start server", error)
        return@execute
      }

      synchronized(stateLock) {
        if (destroyed || serverSocket != null) {
          server.close()
          promise.reject("ERR_SERVER_START", "The server cannot be started", null)
          return@execute
        }
        serverSocket = server
      }
      executor.execute { acceptConnections(server) }
      registerService(server, options, promise)
    }
  }

  private fun registerService(server: ServerSocket, options: ServerOptions, promise: Promise) {
    val listener = object : NsdManager.RegistrationListener {
      override fun onServiceRegistered(serviceInfo: NsdServiceInfo) {
        promise.resolve(
          mapOf(
            "port" to server.localPort,
            "serviceName" to serviceInfo.serviceName,
            "serviceType" to options.serviceType
          )
        )
      }

      override fun onRegistrationFailed(serviceInfo: NsdServiceInfo, errorCode: Int) {
        synchronized(stateLock) { registrationListener = null }
        releaseMulticast()
        closeServerSocket(server)
        promise.reject("ERR_SERVICE_REGISTRATION", "NSD registration failed ($errorCode)", null)
      }

      override fun onServiceUnregistered(serviceInfo: NsdServiceInfo) {
        releaseMulticast()
        val stopPromise = synchronized(stateLock) {
          val pending = serverStopPromise
          serverStopPromise = null
          pending
        }
        stopPromise?.resolve()
      }

      override fun onUnregistrationFailed(serviceInfo: NsdServiceInfo, errorCode: Int) {
        releaseMulticast()
        val stopPromise = synchronized(stateLock) {
          val pending = serverStopPromise
          serverStopPromise = null
          pending
        }
        if (stopPromise != null) {
          stopPromise.reject("ERR_SERVICE_UNREGISTRATION", "NSD unregistration failed ($errorCode)", null)
        } else {
          emitError("ERR_SERVICE_UNREGISTRATION", "NSD unregistration failed ($errorCode)", "server")
        }
      }
    }

    synchronized(stateLock) { registrationListener = listener }
    acquireMulticast()
    val serviceInfo = NsdServiceInfo().apply {
      serviceName = options.serviceName
      serviceType = options.serviceType
      port = server.localPort
    }

    try {
      nsdManager.registerService(serviceInfo, NsdManager.PROTOCOL_DNS_SD, listener)
    } catch (error: Exception) {
      synchronized(stateLock) { registrationListener = null }
      releaseMulticast()
      closeServerSocket(server)
      promise.reject("ERR_SERVICE_REGISTRATION", error.message ?: "Could not register service", error)
    }
  }

  private fun stopServer(promise: Promise?) {
    val listener: NsdManager.RegistrationListener?
    val server: ServerSocket?
    synchronized(stateLock) {
      listener = registrationListener
      registrationListener = null
      server = serverSocket
      serverSocket = null
    }

    try {
      server?.close()
    } catch (_: Exception) {
      // Closing an already closed listener is harmless.
    }
    connections.values.filter { it.incoming }.forEach { closeConnection(it, "server_stopped") }

    if (listener == null) {
      promise?.resolve()
      return
    }

    synchronized(stateLock) { serverStopPromise = promise }
    try {
      nsdManager.unregisterService(listener)
    } catch (error: Exception) {
      synchronized(stateLock) { serverStopPromise = null }
      releaseMulticast()
      promise?.reject("ERR_SERVICE_UNREGISTRATION", error.message ?: "Could not unregister service", error)
    }
  }

  private fun acceptConnections(server: ServerSocket) {
    while (!server.isClosed) {
      try {
        val socket = server.accept()
        if (connections.size >= MAX_CONNECTIONS) {
          socket.close()
          continue
        }
        addConnection(socket, true)
      } catch (error: SocketException) {
        if (!server.isClosed) emitError("ERR_SERVER_ACCEPT", error.message ?: "Accept failed", "server")
        return
      } catch (error: Exception) {
        if (!server.isClosed) emitError("ERR_SERVER_ACCEPT", error.message ?: "Accept failed", "server")
      }
    }
  }

  private fun startDiscovery(serviceType: String, promise: Promise) {
    synchronized(stateLock) {
      if (discoveryListener != null) {
        promise.reject("ERR_DISCOVERY_ALREADY_RUNNING", "Service discovery is already running", null)
        return
      }
    }

    val listener = object : NsdManager.DiscoveryListener {
      override fun onDiscoveryStarted(regType: String) {
        promise.resolve()
      }

      override fun onServiceFound(serviceInfo: NsdServiceInfo) {
        val key = serviceKey(serviceInfo)
        val serviceId = discoveredServiceIds.computeIfAbsent(key) { UUID.randomUUID().toString() }
        discoveredServices[serviceId] = serviceInfo
        sendEvent(
          "onServiceFound",
          mapOf(
            "serviceId" to serviceId,
            "name" to serviceInfo.serviceName,
            "type" to serviceInfo.serviceType
          )
        )
      }

      override fun onServiceLost(serviceInfo: NsdServiceInfo) {
        val serviceId = discoveredServiceIds.remove(serviceKey(serviceInfo)) ?: return
        discoveredServices.remove(serviceId)
        sendEvent("onServiceLost", mapOf("serviceId" to serviceId))
      }

      override fun onDiscoveryStopped(serviceType: String) = Unit

      override fun onStartDiscoveryFailed(serviceType: String, errorCode: Int) {
        synchronized(stateLock) { discoveryListener = null }
        releaseMulticast()
        promise.reject("ERR_DISCOVERY_START", "NSD discovery failed ($errorCode)", null)
      }

      override fun onStopDiscoveryFailed(serviceType: String, errorCode: Int) {
        emitError("ERR_DISCOVERY_STOP", "Stopping NSD discovery failed ($errorCode)", "discovery")
      }
    }

    synchronized(stateLock) { discoveryListener = listener }
    acquireMulticast()
    try {
      nsdManager.discoverServices(serviceType, NsdManager.PROTOCOL_DNS_SD, listener)
    } catch (error: Exception) {
      synchronized(stateLock) { discoveryListener = null }
      releaseMulticast()
      promise.reject("ERR_DISCOVERY_START", error.message ?: "Could not start discovery", error)
    }
  }

  private fun stopDiscovery(promise: Promise?) {
    val listener = synchronized(stateLock) {
      val current = discoveryListener
      discoveryListener = null
      current
    }
    discoveredServices.clear()
    discoveredServiceIds.clear()
    if (listener == null) {
      promise?.resolve()
      return
    }

    try {
      nsdManager.stopServiceDiscovery(listener)
      releaseMulticast()
      promise?.resolve()
    } catch (error: Exception) {
      releaseMulticast()
      promise?.reject("ERR_DISCOVERY_STOP", error.message ?: "Could not stop discovery", error)
    }
  }

  @Suppress("DEPRECATION")
  private fun connectToService(serviceId: String, promise: Promise) {
    val service = discoveredServices[serviceId]
    if (service == null) {
      promise.reject("ERR_SERVICE_NOT_FOUND", "Service '$serviceId' was not found", null)
      return
    }

    try {
      nsdManager.resolveService(service, object : NsdManager.ResolveListener {
        override fun onServiceResolved(serviceInfo: NsdServiceInfo) {
          val address = serviceInfo.host
          if (address == null) {
            promise.reject("ERR_SERVICE_RESOLVE", "The service has no reachable address", null)
            return
          }
          executor.execute { connect(address, serviceInfo.port, promise) }
        }

        override fun onResolveFailed(serviceInfo: NsdServiceInfo, errorCode: Int) {
          promise.reject("ERR_SERVICE_RESOLVE", "NSD resolution failed ($errorCode)", null)
        }
      })
    } catch (error: Exception) {
      promise.reject("ERR_SERVICE_RESOLVE", error.message ?: "Could not resolve service", error)
    }
  }

  private fun connect(host: String, port: Int, promise: Promise) {
    val address = try {
      InetAddress.getByName(host)
    } catch (error: Exception) {
      promise.reject("ERR_CONNECTION_FAILED", error.message ?: "Could not resolve host", error)
      return
    }
    connect(address, port, promise)
  }

  private fun connect(address: InetAddress, port: Int, promise: Promise) {
    if (port !in 1..65535) {
      promise.reject("ERR_INVALID_PORT", "Port must be between 1 and 65535", null)
      return
    }
    val socket = Socket()
    try {
      socket.connect(InetSocketAddress(address, port), CONNECT_TIMEOUT_MS)
      val connection = addConnection(socket, false)
      promise.resolve(connectionInfo(connection))
    } catch (error: Exception) {
      try {
        socket.close()
      } catch (_: Exception) {
        // Ignore cleanup errors after a failed connection.
      }
      promise.reject("ERR_CONNECTION_FAILED", error.message ?: "Could not connect", error)
    }
  }

  private fun addConnection(socket: Socket, incoming: Boolean): ManagedConnection {
    socket.tcpNoDelay = true
    val connection = ManagedConnection(UUID.randomUUID().toString(), socket, incoming)
    connections[connection.id] = connection
    sendEvent("onConnectionOpened", connectionInfo(connection))
    executor.execute { readConnection(connection) }
    return connection
  }

  private fun readConnection(connection: ManagedConnection) {
    val buffer = ByteArray(8 * 1024)
    try {
      val input = connection.socket.getInputStream()
      while (!connection.closed.get()) {
        val count = input.read(buffer)
        if (count < 0) break
        if (count > 0) {
          sendEvent(
            "onMessage",
            mapOf("connectionId" to connection.id, "data" to buffer.copyOf(count))
          )
        }
      }
      closeConnection(connection, "remote_close")
    } catch (error: Exception) {
      if (!connection.closed.get()) closeConnection(connection, "connection_error", error.message)
    }
  }

  private fun write(connection: ManagedConnection, data: ByteArray) {
    try {
      synchronized(connection.writeLock) {
        connection.socket.getOutputStream().apply {
          write(data)
          flush()
        }
      }
    } catch (error: Exception) {
      closeConnection(connection, "connection_error", error.message)
      throw LanSocketsException("ERR_SEND_FAILED", error.message ?: "Could not send data", error)
    }
  }

  private fun closeConnection(connection: ManagedConnection, reason: String, message: String? = null) {
    if (!connection.closed.compareAndSet(false, true)) return
    connections.remove(connection.id)
    try {
      connection.socket.close()
    } catch (_: Exception) {
      // The close event is still valid if the underlying socket was already closed.
    }
    if (!destroyed) {
      sendEvent(
        "onConnectionClosed",
        mapOf("connectionId" to connection.id, "reason" to reason, "message" to message)
      )
    }
  }

  private fun closeServerSocket(server: ServerSocket) {
    synchronized(stateLock) {
      if (serverSocket === server) serverSocket = null
    }
    try {
      server.close()
    } catch (_: Exception) {
      // Ignore cleanup errors while unwinding startup.
    }
  }

  private fun connectionInfo(connection: ManagedConnection): Map<String, Any?> = mapOf(
    "connectionId" to connection.id,
    "remoteAddress" to connection.socket.inetAddress?.hostAddress,
    "remotePort" to connection.socket.port,
    "incoming" to connection.incoming
  )

  private fun serviceKey(serviceInfo: NsdServiceInfo) =
    "${serviceInfo.serviceName}\u0000${serviceInfo.serviceType}"

  private fun acquireMulticast() {
    synchronized(stateLock) {
      multicastUsers += 1
      if (multicastLock?.isHeld == true) return
      val context = appContext.reactContext?.applicationContext ?: return
      val wifiManager = context.getSystemService(Context.WIFI_SERVICE) as WifiManager
      multicastLock = wifiManager.createMulticastLock("ExpoLanSockets").apply {
        setReferenceCounted(false)
        acquire()
      }
    }
  }

  private fun releaseMulticast() {
    synchronized(stateLock) {
      multicastUsers = (multicastUsers - 1).coerceAtLeast(0)
      if (multicastUsers != 0) return
      try {
        if (multicastLock?.isHeld == true) multicastLock?.release()
      } catch (_: Exception) {
        // A lock may already be released during application teardown.
      }
      multicastLock = null
    }
  }

  private fun emitError(code: String, message: String, operation: String) {
    if (!destroyed) sendEvent("onError", mapOf("code" to code, "message" to message, "operation" to operation))
  }
}

private class LanSocketsException(code: String, message: String, cause: Throwable? = null) :
  expo.modules.kotlin.exception.CodedException(code, message, cause)

File: expo-lan-sockets/ios/ExpoLanSocketsModule.swift
import ExpoModulesCore
import Network

private let maxConnections = 32
private let connectTimeout: TimeInterval = 10

private struct ServerOptions: Record {
  @Field
  var serviceName: String = "Expo LAN Game"

  @Field
  var serviceType: String = "_expo-lan-game._tcp."
}

private final class ManagedConnection {
  let id: String
  let connection: NWConnection
  let incoming: Bool
  var opened = false
  var closed = false
  var receiving = false
  var connectPromise: Promise?
  var timeoutWorkItem: DispatchWorkItem?

  init(id: String, connection: NWConnection, incoming: Bool, promise: Promise?) {
    self.id = id
    self.connection = connection
    self.incoming = incoming
    self.connectPromise = promise
  }
}

public final class ExpoLanSocketsModule: Module {
  private let networkQueue = DispatchQueue(label: "expo.lan.sockets.network")
  private var listener: NWListener?
  private var browser: NWBrowser?
  private var connections: [String: ManagedConnection] = [:]
  private var discoveredServices: [String: NWEndpoint] = [:]
  private var discoveredServiceIds: [String: String] = [:]
  private var destroyed = false

  public func definition() -> ModuleDefinition {
    Name("ExpoLanSockets")

    Events(
      "onServiceFound",
      "onServiceLost",
      "onConnectionOpened",
      "onMessage",
      "onConnectionClosed",
      "onError"
    )

    Constant("capabilities") {
      return [
        "tcpServer": true,
        "tcpClient": true,
        "serviceDiscovery": true
      ]
    }

    AsyncFunction("startServerAsync") { (options: ServerOptions, promise: Promise) in
      self.startServer(options: options, promise: promise)
    }

    AsyncFunction("stopServerAsync") { (promise: Promise) in
      self.networkQueue.async {
        self.stopServer()
        promise.resolve()
      }
    }

    AsyncFunction("startDiscoveryAsync") { (serviceType: String, promise: Promise) in
      self.startDiscovery(serviceType: serviceType, promise: promise)
    }

    AsyncFunction("stopDiscoveryAsync") {
      self.stopDiscovery()
    }
    .runOnQueue(networkQueue)

    AsyncFunction("connectAsync") { (host: String, port: Int, promise: Promise) in
      self.networkQueue.async {
        guard let nwPort = NWEndpoint.Port(rawValue: UInt16(exactly: port) ?? 0), nwPort.rawValue > 0 else {
          promise.reject("ERR_INVALID_PORT", "Port must be between 1 and 65535")
          return
        }
        self.connect(endpoint: .hostPort(host: NWEndpoint.Host(host), port: nwPort), promise: promise)
      }
    }

    AsyncFunction("connectToServiceAsync") { (serviceId: String, promise: Promise) in
      self.networkQueue.async {
        guard let endpoint = self.discoveredServices[serviceId] else {
          promise.reject("ERR_SERVICE_NOT_FOUND", "Service '\(serviceId)' was not found")
          return
        }
        self.connect(endpoint: endpoint, promise: promise)
      }
    }

    AsyncFunction("sendAsync") { (connectionId: String, data: Data, promise: Promise) in
      self.networkQueue.async {
        self.send(connectionId: connectionId, data: data, promise: promise)
      }
    }

    AsyncFunction("broadcastAsync") { (data: Data, promise: Promise) in
      self.networkQueue.async {
        self.broadcast(data: data, promise: promise)
      }
    }

    AsyncFunction("disconnectAsync") { (connectionId: String, promise: Promise) in
      self.networkQueue.async {
        if let connection = self.connections[connectionId] {
          self.closeConnection(connection, reason: "local_close")
        }
        promise.resolve()
      }
    }

    OnDestroy {
      self.networkQueue.sync {
        self.destroyed = true
        self.stopDiscovery()
        self.stopServer()
        Array(self.connections.values).forEach { connection in
          self.closeConnection(connection, reason: "local_close", emit: false)
        }
      }
    }
  }

  private func startServer(options: ServerOptions, promise: Promise) {
    networkQueue.async {
      guard self.listener == nil else {
        promise.reject("ERR_SERVER_ALREADY_RUNNING", "A server is already running")
        return
      }
      guard !self.destroyed else {
        promise.reject("ERR_SERVER_START", "The server cannot be started")
        return
      }

      do {
        let parameters = self.tcpParameters()
        let listener = try NWListener(using: parameters, on: .any)
        var didSettleStart = false
        listener.service = NWListener.Service(
          name: options.serviceName,
          type: self.normalizedServiceType(options.serviceType)
        )
        self.listener = listener

        listener.newConnectionHandler = { [weak self, weak listener] connection in
          guard let self, let listener, self.listener === listener else {
            connection.cancel()
            return
          }
          self.networkQueue.async {
            guard self.listener === listener, self.connections.count < maxConnections else {
              connection.cancel()
              return
            }
            self.addConnection(connection, incoming: true, promise: nil)
          }
        }

        listener.stateUpdateHandler = { [weak self, weak listener] state in
          guard let self, let listener else { return }
          self.networkQueue.async {
            guard self.listener === listener else { return }
            switch state {
            case .ready:
              guard !didSettleStart else { return }
              didSettleStart = true
              guard let port = listener.port?.rawValue else {
                self.listener = nil
                listener.cancel()
                promise.reject("ERR_SERVER_START", "The TCP listener has no port")
                return
              }
              promise.resolve([
                "port": Int(port),
                "serviceName": options.serviceName,
                "serviceType": options.serviceType
              ])
            case .failed(let error):
              self.listener = nil
              listener.cancel()
              self.closeIncomingConnections(reason: "server_stopped")
              if didSettleStart {
                self.emitError(code: "ERR_SERVER_FAILED", message: error.localizedDescription, operation: "server")
              } else {
                didSettleStart = true
                promise.reject("ERR_SERVER_START", error.localizedDescription)
              }
            case .waiting(let error):
              self.listener = nil
              listener.cancel()
              if didSettleStart {
                self.emitError(code: "ERR_SERVER_FAILED", message: error.localizedDescription, operation: "server")
              } else {
                didSettleStart = true
                promise.reject("ERR_SERVER_START", error.localizedDescription)
              }
            default:
              break
            }
          }
        }
        listener.start(queue: self.networkQueue)
      } catch {
        self.listener = nil
        promise.reject("ERR_SERVER_START", error.localizedDescription)
      }
    }
  }

  private func stopServer() {
    listener?.stateUpdateHandler = nil
    listener?.newConnectionHandler = nil
    listener?.cancel()
    listener = nil
    closeIncomingConnections(reason: "server_stopped")
  }

  private func closeIncomingConnections(reason: String) {
    connections.values.filter(\.incoming).forEach { connection in
      closeConnection(connection, reason: reason)
    }
  }

  private func startDiscovery(serviceType: String, promise: Promise) {
    networkQueue.async {
      guard !self.destroyed else {
        promise.reject("ERR_DISCOVERY_START", "Service discovery cannot be started")
        return
      }
      if self.browser != nil {
        self.stopDiscovery()
      }

      let browser = NWBrowser(
        for: .bonjour(type: self.normalizedServiceType(serviceType), domain: nil),
        using: self.tcpParameters()
      )
      var didSettleStart = false
      self.browser = browser
      browser.browseResultsChangedHandler = { [weak self, weak browser] results, _ in
        guard let self, let browser else { return }
        self.networkQueue.async {
          guard self.browser === browser else { return }
          self.updateDiscoveredServices(results)
        }
      }
      browser.stateUpdateHandler = { [weak self, weak browser] state in
        guard let self, let browser else { return }
        self.networkQueue.async {
          guard self.browser === browser else { return }
          switch state {
          case .ready:
            if !didSettleStart {
              didSettleStart = true
              promise.resolve()
            }
          case .failed(let error):
            self.stopDiscovery()
            if didSettleStart {
              self.emitError(code: "ERR_DISCOVERY_FAILED", message: error.localizedDescription, operation: "discovery")
            } else {
              didSettleStart = true
              promise.reject("ERR_DISCOVERY_START", error.localizedDescription)
            }
          case .waiting(let error):
            self.stopDiscovery()
            if didSettleStart {
              self.emitError(code: "ERR_DISCOVERY_FAILED", message: error.localizedDescription, operation: "discovery")
            } else {
              didSettleStart = true
              promise.reject("ERR_DISCOVERY_START", error.localizedDescription)
            }
          default:
            break
          }
        }
      }
      browser.start(queue: self.networkQueue)
    }
  }

  private func stopDiscovery() {
    browser?.stateUpdateHandler = nil
    browser?.browseResultsChangedHandler = nil
    browser?.cancel()
    browser = nil
    discoveredServices.removeAll()
    discoveredServiceIds.removeAll()
  }

  private func updateDiscoveredServices(_ results: Set<NWBrowser.Result>) {
    var activeKeys = Set<String>()
    for result in results {
      guard let service = serviceDetails(result.endpoint) else { continue }
      activeKeys.insert(service.key)
      let serviceId = discoveredServiceIds[service.key] ?? UUID().uuidString
      let isNew = discoveredServiceIds[service.key] == nil
      discoveredServiceIds[service.key] = serviceId
      discoveredServices[serviceId] = result.endpoint
      if isNew {
        sendEvent("onServiceFound", [
          "serviceId": serviceId,
          "name": service.name,
          "type": service.type
        ])
      }
    }

    let lostKeys = discoveredServiceIds.keys.filter { !activeKeys.contains($0) }
    for key in lostKeys {
      guard let serviceId = discoveredServiceIds.removeValue(forKey: key) else { continue }
      discoveredServices.removeValue(forKey: serviceId)
      sendEvent("onServiceLost", ["serviceId": serviceId])
    }
  }

  private func connect(endpoint: NWEndpoint, promise: Promise) {
    guard !destroyed else {
      promise.reject("ERR_CONNECTION_FAILED", "The socket module is closed")
      return
    }
    let connection = NWConnection(to: endpoint, using: tcpParameters())
    let managed = addConnection(connection, incoming: false, promise: promise)
    let timeout = DispatchWorkItem { [weak self, weak managed] in
      guard let self, let managed, !managed.opened, !managed.closed else { return }
      managed.connectPromise?.reject("ERR_CONNECTION_FAILED", "Connection timed out")
      managed.connectPromise = nil
      self.closeConnection(managed, reason: "connection_error", emit: false)
    }
    managed.timeoutWorkItem = timeout
    networkQueue.asyncAfter(deadline: .now() + connectTimeout, execute: timeout)
  }

  @discardableResult
  private func addConnection(_ connection: NWConnection, incoming: Bool, promise: Promise?) -> ManagedConnection {
    let managed = ManagedConnection(id: UUID().uuidString, connection: connection, incoming: incoming, promise: promise)
    connections[managed.id] = managed
    connection.stateUpdateHandler = { [weak self, weak managed] state in
      guard let self, let managed else { return }
      self.networkQueue.async {
        guard self.connections[managed.id] === managed, !managed.closed else { return }
        switch state {
        case .ready:
          guard !managed.opened else { return }
          managed.opened = true
          managed.timeoutWorkItem?.cancel()
          managed.timeoutWorkItem = nil
          let info = self.connectionInfo(managed)
          self.sendEvent("onConnectionOpened", info)
          managed.connectPromise?.resolve(info)
          managed.connectPromise = nil
          self.receive(managed)
        case .failed(let error):
          let message = self.networkErrorMessage(error, endpoint: managed.connection.endpoint)
          managed.connectPromise?.reject("ERR_CONNECTION_FAILED", message)
          managed.connectPromise = nil
          self.closeConnection(
            managed,
            reason: "connection_error",
            message: message,
            emit: managed.opened
          )
        case .cancelled:
          self.closeConnection(managed, reason: "remote_close", emit: managed.opened)
        default:
          break
        }
      }
    }
    connection.start(queue: networkQueue)
    return managed
  }

  private func receive(_ managed: ManagedConnection) {
    guard !managed.receiving, !managed.closed else { return }
    managed.receiving = true

    func receiveNext() {
      managed.connection.receive(minimumIncompleteLength: 1, maximumLength: 8 * 1024) { [weak self, weak managed] data, _, isComplete, error in
        guard let self, let managed else { return }
        self.networkQueue.async {
          guard self.connections[managed.id] === managed, !managed.closed else { return }
          if let data, !data.isEmpty {
            self.sendEvent("onMessage", ["connectionId": managed.id, "data": data])
          }
          if let error {
            self.closeConnection(managed, reason: "connection_error", message: error.localizedDescription)
          } else if isComplete {
            self.closeConnection(managed, reason: "remote_close")
          } else {
            receiveNext()
          }
        }
      }
    }

    receiveNext()
  }

  private func send(connectionId: String, data: Data, promise: Promise) {
    guard let managed = connections[connectionId], managed.opened, !managed.closed else {
      promise.reject("ERR_CONNECTION_NOT_FOUND", "Connection '\(connectionId)' was not found")
      return
    }
    managed.connection.send(content: data, completion: .contentProcessed { [weak self, weak managed] error in
      guard let self, let managed else { return }
      self.networkQueue.async {
        if let error {
          self.closeConnection(managed, reason: "connection_error", message: error.localizedDescription)
          promise.reject("ERR_SEND_FAILED", error.localizedDescription)
        } else {
          promise.resolve()
        }
      }
    })
  }

  private func broadcast(data: Data, promise: Promise) {
    let activeConnections = connections.values.filter { $0.opened && !$0.closed }
    guard !activeConnections.isEmpty else {
      promise.resolve()
      return
    }

    var remaining = activeConnections.count
    var firstError: NWError?
    for managed in activeConnections {
      managed.connection.send(content: data, completion: .contentProcessed { [weak self, weak managed] error in
        guard let self else { return }
        self.networkQueue.async {
          if let error {
            firstError = firstError ?? error
            if let managed {
              self.closeConnection(managed, reason: "connection_error", message: error.localizedDescription)
            }
          }
          remaining -= 1
          if remaining == 0 {
            if let firstError {
              promise.reject("ERR_SEND_FAILED", firstError.localizedDescription)
            } else {
              promise.resolve()
            }
          }
        }
      })
    }
  }

  private func closeConnection(
    _ managed: ManagedConnection,
    reason: String,
    message: String? = nil,
    emit: Bool = true
  ) {
    guard connections[managed.id] === managed, !managed.closed else { return }
    managed.closed = true
    managed.timeoutWorkItem?.cancel()
    managed.timeoutWorkItem = nil
    connections.removeValue(forKey: managed.id)
    managed.connection.stateUpdateHandler = nil
    managed.connection.cancel()
    if emit && !destroyed {
      let payload: [String: Any?] = [
        "connectionId": managed.id,
        "reason": reason,
        "message": message
      ]
      sendEvent("onConnectionClosed", payload)
    }
  }

  private func connectionInfo(_ managed: ManagedConnection) -> [String: Any?] {
    var address: String?
    var port = 0
    switch managed.connection.endpoint {
    case .hostPort(let host, let endpointPort):
      address = "\(host)"
      port = Int(endpointPort.rawValue)
    case .service(let name, _, _, _):
      address = name
    default:
      break
    }
    return [
      "connectionId": managed.id,
      "remoteAddress": address,
      "remotePort": port,
      "incoming": managed.incoming
    ]
  }

  private func serviceDetails(_ endpoint: NWEndpoint) -> (key: String, name: String, type: String)? {
    guard case .service(let name, let type, let domain, let interface) = endpoint else { return nil }
    let key = "\(name)\u{0}\(type)\u{0}\(domain)\u{0}\(String(describing: interface))"
    return (key, name, type.hasSuffix(".") ? type : "\(type).")
  }

  private func normalizedServiceType(_ serviceType: String) -> String {
    serviceType.hasSuffix(".") ? String(serviceType.dropLast()) : serviceType
  }

  private func networkErrorMessage(_ error: NWError, endpoint: NWEndpoint) -> String {
    let reason: String
    switch error {
    case .posix(let code):
      reason = "POSIX \(code.rawValue) (\(code))"
    case .dns(let code):
      reason = "DNS \(code)"
    case .tls(let status):
      reason = "TLS \(status)"
    case .wifiAware(let error):
      reason = "Wi-Fi Aware \(error)"
    @unknown default:
      reason = String(describing: error)
    }
    return "Could not connect to \(endpoint): \(reason)"
  }

  private func tcpParameters() -> NWParameters {
    let parameters = NWParameters.tcp
    parameters.includePeerToPeer = true
    parameters.allowLocalEndpointReuse = true
    return parameters
  }

  private func emitError(code: String, message: String, operation: String) {
    guard !destroyed else { return }
    sendEvent("onError", ["code": code, "message": message, "operation": operation])
  }
}
